# eBay Car Sales

In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The original dataset can be found in [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data), which contains 370,000+ used cars info, but for the purposes of this project we'll use a sample of 50,000.

## First look at the data

We'll start by importing our `Numpy` and `Pandas` libraries to handle the data set more easily, and then reading the csv file to load all the data available.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv',encoding='Latin-1')

Now that we've loaded our dataset, we'll take a quick look at the first rows of our dataframe and some information about it.

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

As we can see our data set is described by 20 columns, and is mainly composed of integers and strings. On the other hand we can see how some columns like `vehicleType` or `gearbox` have some missing values, we'll take a closer look at this.

In [4]:
for col in autos.columns:
    nulls_mask = autos.loc[:,col].isnull()
    num_nulls = autos.loc[nulls_mask,col].shape[0]
    if num_nulls != 0:
        percentage = round((num_nulls/autos.shape[0])*100,2)
        print(col,'column has',num_nulls,'missing values, corresponding to the',str(percentage)+'%','of total records')

vehicleType column has 5095 missing values, corresponding to the 10.19% of total records
gearbox column has 2680 missing values, corresponding to the 5.36% of total records
model column has 2758 missing values, corresponding to the 5.52% of total records
fuelType column has 4482 missing values, corresponding to the 8.96% of total records
notRepairedDamage column has 9829 missing values, corresponding to the 19.66% of total records


## Cleaning the data

Now let's take a quick look at our column names.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll change some column names and use [Snake case](https://en.wikipedia.org/wiki/Snake_case) instead of [Camel case](https://en.wikipedia.org/wiki/Camel_case) to give it a more clean look.

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now let's look at some descriptive statistics to see which columns are useful and which aren't, in order to concentrate our efforts in analyzing the right data.

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The first thing we can observe is how the `seller` and `offer_type` columns contain 1 value for almost all the cases (49,999 out of 50,000), so they're strong candidates to be dropped. On the other hand we see `price` and `odomoeter` are numeric values but they're stored as text, so we'll start by correcting that.

In [8]:
def clean_text(text,symbol):
    text = text.replace(symbol,'')
    return text

autos['price'] = [clean_text(row,'$') for row in autos['price']]
autos['price'] = [clean_text(row,',') for row in autos['price']]
autos['price'] = autos['price'].astype(dtype='int64')
autos['odometer'] = [clean_text(row,'km') for row in autos['odometer']]
autos['odometer'] = [clean_text(row,',') for row in autos['odometer']]
autos['odometer'] = autos['odometer'].astype(dtype='int64')
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)

Now let's take an even closer look to the `odometer_km` and `price` columns to see if there's any outlier or anything that doesn't look right.

In [9]:
autos['odometer_km'].unique().shape[0]

13

In [10]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [11]:
autos['odometer_km'].value_counts().head(15)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The `odometer_km` column seems to be fine, given that there are 13 unique values, we can infer the values actually represent ranges of kilometers and the last value also holds the 150,000+ instances. Now let's check the `price` column.

In [12]:
autos['price'].unique().shape[0]

2357

In [13]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [14]:
autos['price'].value_counts().sort_index(ascending=True).head(15)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64

In [15]:
autos['price'].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

Now we see some unusual values. First we observe cars with very low prices (like 0 or 1), which don't correspond to a car's value, so we'll drop these. On the other hand, considering we have a mean of 9,840 it's very odd the standard deviation is 481,104, this must be due to some outliers in the highest prices, so we'll drop these too.

In [16]:
autos = autos[autos['price'].between(100,350000)]

We now observe below some more realistic statistics.

In [17]:
autos['price'].describe()

count     48224.000000
mean       5930.371433
std        9078.372762
min         100.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

## Analyzing the data - Part 1

Moving on we'll continue by analyzing the date columns in our dataset, these are `date_crawled`,`last_seen`,`ad_created`,`registration_month` and `registration_year`. Let's look at the first three of these which correspond to datetime format.

In [18]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We can see how the first 10 characters represent the date of each column, while the rest of the characters is the time. We'll analyze the dates for each of these 3 columns.

In [19]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025361
2016-03-06    0.014039
2016-03-07    0.036061
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-10    0.032287
2016-03-11    0.032598
2016-03-12    0.036911
2016-03-13    0.015677
2016-03-14    0.036662
2016-03-15    0.034319
2016-03-16    0.029467
2016-03-17    0.031499
2016-03-18    0.012898
2016-03-19    0.034734
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-22    0.032888
2016-03-23    0.032287
2016-03-24    0.029446
2016-03-25    0.031499
2016-03-26    0.032308
2016-03-27    0.031126
2016-03-28    0.034962
2016-03-29    0.034112
2016-03-30    0.033738
2016-03-31    0.031851
2016-04-01    0.033697
2016-04-02    0.035605
2016-04-03    0.038611
2016-04-04    0.036538
2016-04-05    0.013064
2016-04-06    0.003173
2016-04-07    0.001389
Name: date_crawled, dtype: float64

For the `date_crawled` column we have 34 unique dates, for each day from the 2016-03-05 until the 2016-04-07, with values quite uniformly distributed, ranging between 0 and 4%.

In [20]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033096
2016-03-10    0.031997
2016-03-11    0.032909
2016-03-12    0.036745
2016-03-13    0.017045
2016-03-14    0.035294
2016-03-15    0.034049
2016-03-16    0.029964
2016-03-17    0.031167
2016-03-18    0.013582
2016-03-19    0.033614
2016-03-20    0.037865
2016-03-21 

For the `ad_created` column we have 76 unique dates, but here we observe how the newest dates have greater percentages, which is logic, cars posted a while ago should've been already sold while the newer ones are still online.

In [21]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001078
2016-03-06    0.004313
2016-03-07    0.005433
2016-03-08    0.007320
2016-03-09    0.009580
2016-03-10    0.010638
2016-03-11    0.012400
2016-03-12    0.023785
2016-03-13    0.008875
2016-03-14    0.012629
2016-03-15    0.015863
2016-03-16    0.016444
2016-03-17    0.028098
2016-03-18    0.007320
2016-03-19    0.015760
2016-03-20    0.020654
2016-03-21    0.020550
2016-03-22    0.021359
2016-03-23    0.018580
2016-03-24    0.019762
2016-03-25    0.019098
2016-03-26    0.016672
2016-03-27    0.015552
2016-03-28    0.020840
2016-03-29    0.022292
2016-03-30    0.024697
2016-03-31    0.023826
2016-04-01    0.022852
2016-04-02    0.024884
2016-04-03    0.025133
2016-04-04    0.024531
2016-04-05    0.125062
2016-04-06    0.221964
2016-04-07    0.132154
Name: last_seen, dtype: float64

For the `last_seen` column we have 34 unique dates as well as `date_crawled` because they represent the exact same dates. In this case though, we see a clear trend where the last 3 days have very high values (12, 22 and 13%) and the days before start to oscillate around 2% and slowly decrease.

In [22]:
autos['registration_year'].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

For the `registration_year` there are some clear outliers like 1000 (minimum) and 9999 (maximum) which corrupts our analysis, but besides that with the quartile information we can see every fourth of cars published is registered every 4 years. Now we'll filter our dataset in order to have only possible years for the registration of a car, and see the effects of this.

In [23]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True).sort_index(ascending=False)

2016    0.025932
2015    0.008198
2014    0.014282
2013    0.017281
2012    0.028219
2011    0.034907
2010    0.034238
2009    0.044874
2008    0.047679
2007    0.049038
2006    0.057560
2005    0.062802
2004    0.058228
2003    0.058099
2002    0.053439
2001    0.056718
2000    0.066966
1999    0.062112
1998    0.050483
1997    0.041530
1996    0.029233
1995    0.025738
1994    0.013505
1993    0.009061
1992    0.007918
1991    0.007292
1990    0.007163
1989    0.003689
1988    0.002869
1987    0.001553
          ...   
1968    0.000561
1967    0.000561
1966    0.000475
1965    0.000367
1964    0.000259
1963    0.000173
1962    0.000086
1961    0.000129
1960    0.000475
1959    0.000129
1958    0.000086
1957    0.000043
1956    0.000086
1955    0.000043
1954    0.000043
1953    0.000022
1952    0.000022
1951    0.000043
1950    0.000022
1948    0.000022
1943    0.000022
1941    0.000043
1939    0.000022
1938    0.000022
1937    0.000086
1934    0.000043
1931    0.000022
1929    0.0000

Now the data makes more sense since we're working with the right years. Logically newer cars are more abundant and older cars are more rare.

## Analyzing the data - Part 2

Now we'll understand some more about the `brand` column, for this we'll be working with aggregated data in order to know the top models and vehicle types by brand, and the average selling price by brand, all of this for the top 10 brands. So first we'll start by selecting the top 10 brands and then work with them.

In [52]:
brands = autos['brand'].value_counts().keys().tolist()[:10]

In [53]:
top_model_by_brand = {}

for b in brands:
    selected_rows = autos[autos['brand'] == b]
    sorted_by_model = selected_rows.sort_values(by='model',ascending=False)
    first_row = sorted_by_model.iloc[0,:]
    model = first_row.loc['model']
    top_model_by_brand[b] = model
    
for key in top_model_by_brand:
    print((key.title())+'\'s','top model is',str(top_model_by_brand[key]).title())

Volkswagen's top model is Up
Bmw's top model is Z_Reihe
Opel's top model is Zafira
Mercedes_Benz's top model is Vito
Audi's top model is Tt
Ford's top model is Transit
Renault's top model is Twingo
Peugeot's top model is Andere
Fiat's top model is Stilo
Seat's top model is Toledo


In [54]:
top_type_by_brand = {}

for b in brands:
    selected_rows = autos[autos['brand'] == b]
    sorted_by_type = selected_rows.sort_values(by='vehicle_type',ascending=False)
    first_row = sorted_by_type.iloc[0,:]
    vtype = first_row.loc['vehicle_type']
    top_type_by_brand[b] = vtype
    
for key in top_type_by_brand:
    print((key.title())+'\'s','top vehicle type is',top_type_by_brand[key].title())

Volkswagen's top vehicle type is Suv
Bmw's top vehicle type is Suv
Opel's top vehicle type is Suv
Mercedes_Benz's top vehicle type is Suv
Audi's top vehicle type is Suv
Ford's top vehicle type is Suv
Renault's top vehicle type is Suv
Peugeot's top vehicle type is Suv
Fiat's top vehicle type is Suv
Seat's top vehicle type is Limousine


In [55]:
avg_price_by_brand = {}

for b in brands:
    selected_rows = autos[autos['brand'] == b]
    prices = selected_rows['price']
    avg_price = prices.mean()
    avg_price_by_brand[b] = avg_price
    
for key in avg_price_by_brand:
    print((key.title())+'\'s','average price is','$'+str(round(avg_price_by_brand[key],2)))

Volkswagen's average price is $5436.95
Bmw's average price is $8381.68
Opel's average price is $3005.5
Mercedes_Benz's average price is $8672.65
Audi's average price is $9380.72
Ford's average price is $3779.27
Renault's average price is $2496.07
Peugeot's average price is $3113.86
Fiat's average price is $2836.87
Seat's average price is $4433.42


In [59]:
avg_mileage_by_brand = {}

for b in brands:
    selected_rows = autos[autos['brand'] == b]
    mileage = selected_rows['odometer_km']
    avg_mileage = mileage.mean()
    avg_mileage_by_brand[b] = avg_mileage
    
for key in avg_mileage_by_brand:
    print((key.title())+'\'s','average mileage is',str(round(avg_mileage_by_brand[key],2)))

Volkswagen's average mileage is 128799.88
Bmw's average mileage is 132695.32
Opel's average mileage is 129384.43
Mercedes_Benz's average mileage is 131025.67
Audi's average mileage is 129245.4
Ford's average mileage is 124277.11
Renault's average mileage is 128281.39
Peugeot's average mileage is 127127.89
Fiat's average mileage is 116950.29
Seat's average mileage is 121536.64


The first obvious observations is how SUVs dominate the german market, with 9 out of 10 of the top brands ruled by this type of vehicle. Then in relation to the price we have some very cheap brands like Fiat or Peugeot, and then some others that cost in average almost three times, like Audi and Mercedes Benz. Finally, all brands have similar average mileages, which are around 125,000-130,000.